airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/up_to_date/pipeline.py

```diff
@@ -0,0 +1,366 @@
+
+    title = "Check if the connector is a candidate for updating."
+
+    def __init__(self, context: PipelineContext) -> None:
+        super().__init__(context)
+
+    async def _run(self) -> StepResult:
+        connector_dir_entries = await (await self.context.get_connector_dir()).entries()
+        if self.context.connector.language not in [ConnectorLanguage.PYTHON, ConnectorLanguage.LOW_CODE]:
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector is not a Python connector.",
+            )
+        if self.context.connector.connector_type != "source":
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector is not a source connector.",
+            )
+        if POETRY_LOCK_FILE not in connector_dir_entries or POETRY_TOML_FILE not in connector_dir_entries:
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector requires poetry.",
+            )
+
+        if not self.context.connector.metadata or not self.context.connector.metadata.get("connectorBuildOptions", {}).get("baseImage"):
+            return StepResult(
+                step=self,
+                status=StepStatus.SKIPPED,
+                stderr="The connector can't be updated because it does not have a base image defined in the metadata.",
+            )
+
+        return StepResult(
+            step=self,
+            status=StepStatus.SUCCESS,
+        )
+
+
+class UpdatePoetry(Step):
+    context: ConnectorContext
+    dev: bool
+    specified_versions: dict[str, str]
+
+    title = "Update versions of libraries in poetry."
+
+    def __init__(self, context: PipelineContext, dev: bool, specific_dependencies: List[str]) -> None:
+        super().__init__(context)
+        self.dev = dev
+        self.specified_versions = parse_specific_dependencies(specific_dependencies)
+
+    async def _run(self) -> StepResult:
+        base_image_name = self.context.connector.metadata["connectorBuildOptions"]["baseImage"]
+        base_container = self.dagger_client.container(platform=LOCAL_BUILD_PLATFORM).from_(base_image_name)
+        connector_container = await with_python_connector_installed(
+            self.context,
+            base_container,
+            str(self.context.connector.code_directory),
+        )
+
+        try:
+            before_versions = await get_poetry_versions(connector_container)
+            before_main = await get_poetry_versions(connector_container, only="main")
+
+            if self.specified_versions:
+                for package, dep in self.specified_versions.items():
+                    self.logger.info(f"  Specified: poetry add {dep}")
+                    if package in before_main:
+                        connector_container = await connector_container.with_exec(["poetry", "add", dep])
```

Comment:
@alafanechere can you help me here? Does this make a new container each time so we are only ending up with the last one or will this keep building on itself?

Reply:
if it's a new one each time, maybe I need to export the two files and mount?

Reply:
You can consider it's returning a new container on which the `with_exec` will be executed.
Execution is lazy so nothing actually happens until you await the container.

Reply:
> maybe I need to export the two files and mount?

No you don't have to do this dance, you can consider that at the end of the for loop `connector_container` will basically be the `connector_container` from before the `for` to which all the `with_exec` with `poetry add` are appended.